private/buf/cmd/buf/command/beta/plugin/pluginpush/pluginpush.go

```diff
@@ -109,38 +112,46 @@ func run(
 	}
 	// TODO: Once we support multiple plugin source types, this could be abstracted away
 	// in the bufpluginsource package. This is much simpler for now though.
-	pluginSourceFilePath, err := storage.Exists(ctx, sourceBucket, bufpluginsource.DockerSourceFilePath)
+	dockerfile, err := sourceBucket.Get(ctx, bufplugindocker.SourceFilePath)
 	if err != nil {
-		return bufcli.NewInternalError(err)
+		if errors.Is(err, os.ErrNotExist) {
+			return fmt.Errorf("please define a %s plugin source file in the target directory", bufplugindocker.SourceFilePath)
+		}
+		return err
 	}
-	if !pluginSourceFilePath {
-		return fmt.Errorf("please define a %s plugin source file in the target directory", bufpluginsource.DockerSourceFilePath)
+	defer func() {
+		retErr = multierr.Append(retErr, dockerfile.Close())
+	}()
+
+	client, err := bufplugindocker.NewClient(container.Logger())
+	if err != nil {
+		return err
 	}
-	// TODO: Build and push the image to the OCI registry with the Docker library.
-	// Once complete, use the digest in RPC below.
-	//
-	// We effectively need to codify the following:
-	//
-	//  $ docker build -f Dockerfile.twirp -t plugins.buf.build/demolab/twirp:v8.1.0-1 .
-	//  $ docker push plugins.buf.build/demolab/twirp:v8.1.0-1
-	//
-	// This will look something along the lines of:
-	//
-	//  readObjectCloser, err := sourceBucket.Get(pluginSourceFilePath)
-	//  if err != nil {
-	//    return nil, err
-	//  }
-	//  defer func() {
-	//    retErr = multierr.Append(retErr, readObjectCloser.Close())
-	//  }()
-	//  dockerfileContent, err := io.ReadAll(readObjectCloser)
-	//  if err != nil {
-	//    return nil, err
-	//  }
-	//  // Build and push to the OCI registry using the Docker library ...
-	//
-	// ---
-	//
+	defer func() {
+		retErr = multierr.Append(retErr, client.Close())
+	}()
+
+	buildResponse, err := client.Build(ctx, dockerfile, pluginConfig, bufplugindocker.WithConfigDirPath(container.ConfigDirPath()))
+	if err != nil {
+		return err
+	}
+
+	// We build a Docker image using a unique ID label each time.
+	// After we're done publishing the image, we delete it to not leave a lot of images left behind.
+	// buildkit maintains a separate build cache so removing the image doesn't appear to impact future rebuilds.
+	defer func() {
+		if _, err := client.Delete(ctx, buildResponse.Image); err != nil {
+			// Not fatal - we did our best to clean up after ourselves.
+			container.Logger().Warn("failed to delete image", zap.String("image", buildResponse.Image), zap.Error(err))
```

Comment:
We might just want to use `multierr` here and let the command return the error (similar to what we do for `client.Close` above). I know that there's a good reason to emit a warning (the command was technically successful - we just didn't clean up some of the resources), but returning the error in these cases is consistent with our other commands.
```suggestion
			retErr = mutlierr.Append(retErr, fmt.Errorf("failed to delete image %q", buildResponse.Image))
```